tfhe/src/integer/server_key/radix/shift.rs

```diff
@@ -222,15 +267,85 @@ impl ServerKey {
     ///
     /// // Decrypt:
-    /// let dec: u64 = cks.decrypt(&ct);
+    /// let dec = cks.decrypt(&ct);
     /// assert_eq!(msg << shift, dec);
     /// ```
     pub fn unchecked_scalar_left_shift_assign<PBSOrder: PBSOrderMarker>(
         &self,
         ct: &mut RadixCiphertext<PBSOrder>,
-        shift: usize,
+        shift: u64,
     ) {
-        let tmp = 1_u64 << shift;
-        self.smart_scalar_mul_assign(ct, tmp);
+        // see parallel implementation for a bit more details
+
+        debug_assert!(ct.block_carries_are_empty());
+        debug_assert!(self.key.carry_modulus.0 >= self.key.message_modulus.0 / 2);
+
+        if shift == 0 {
+            return;
+        }
+
+        let num_bits_in_block = self.key.message_modulus.0.ilog2() as u64;
+        let rotations = ((shift / num_bits_in_block) as usize).min(ct.blocks.len());
+        let shift_within_block = shift % num_bits_in_block;
+        let total_num_bits = num_bits_in_block * ct.blocks.len() as u64;
+
+        if shift > total_num_bits {
+            for block in ct.blocks.iter_mut() {
+                self.key.create_trivial_assign(block, 0)
+            }
+            return;
+        }
+
+        // rotate right as the blocks are from LSB to MSB
+        ct.blocks.rotate_right(rotations);
+        for block in &mut ct.blocks[..rotations] {
+            self.key.create_trivial_assign(block, 0)
+        }
+
+        if shift_within_block == 0 || rotations == ct.blocks.len() {
+            return;
+        }
+
+        let lut = self
+            .key
+            .generate_accumulator_bivariate(|current_block, previous_block| {
+                let current_block = current_block << shift_within_block;
+                let previous_block = previous_block << shift_within_block;
+
+                let message_of_current_block = current_block % self.key.message_modulus.0 as u64;
+                let carry_of_previous_block = previous_block / self.key.message_modulus.0 as u64;
+                message_of_current_block + carry_of_previous_block
+            });
+        let partial_blocks = ct.blocks[rotations..]
+            .windows(2)
+            .map(|blocks| {
+                // We are right-shifting,
+                // so we get the bits from the next block in the vec
+                let (previous_block, current_block) = (&blocks[0], &blocks[1]);
+                self.key
+                    .unchecked_apply_lookup_table_bivariate(current_block, previous_block, &lut)
+            })
+            .collect::<Vec<_>>();
+
+        // We do this block separately as this one does not
+        // need to get incoming bits from it neighbour
+        self.key.unchecked_scalar_left_shift_assign(
+            &mut ct.blocks[rotations],
+            shift_within_block as u8,
+        );
+        // as left shift is not done via PBS we message extract
+        self.key.message_extract_assign(&mut ct.blocks[rotations]);
```

Comment:
let's perform a single PBS here maybe ?

Reply:
I mean perform shift + message extract in one go

Reply:
yeah we can

Reply:
looks like this one is not fused in a single PBS?